### ISSUE TITLE
Illegal division by zero error in LTR.identifier.pl and “Killed” messages during LTR candidate identification

### ISSUE THREAD

**Issue by kaurharpreet-umn** [oushujun/EDTA#564]
Hello,

I’m running into issues while executing EDTA (v2.2.2). I installed the pipeline using the following mamba command:

mamba create -n EDTA -c conda-forge -c bioconda -c r annosine2 biopython blast cd-hit coreutils genericrepeatfinder genometools-genometools glob2 h5py==3.9 keras==2.11 ltr_finder ltr_retriever mdust multiprocess muscle openjdk pandas perl perl-text-soundex pyarrow python r-base r-dplyr regex repeatmodeler r-ggplot2 r-here r-tidyr scikit-learn swifter tensorflow==2.11 tesorter

After creating the environment, I manually installed additional dependencies:
HelitronScanner
LTR_FINDER_parallel
LTR_HARVEST_parallel
TIR-Learner

EDTA started running fine but crashed after ~2 days of execution. Here are the errors from the log:

Log Excerpt:
Identify LTR retrotransposon candidates from scratch.

sh: line 1: 2939980 Killed                  timeout -s KILL 300s /path/to/env/gt ltrharvest ...
sh: line 1: 3039260 Killed                  timeout -s KILL 300 /path/to/env/ltr_finder ...

Thread 17 terminated abnormally: Illegal division by zero at ... LTR.identifier.pl line 253
Thread 14 terminated abnormally: Illegal division by zero at ... LTR.identifier.pl line 253
... (many similar lines) ...

My command:
perl EDTA.pl --genome all.haps.rename.shortid.fasta \
        --species others \
        --step all \
        --cds medsa.RegenSY27x.gnm1.ann1.gene_models_main.shortid.cds.fasta \
        --overwrite 1 \
        --sensitive 1 \
        --anno 1 \
        --maxdiv 20 \
        --evaluate 1 \
        --threads 90 \
        --force 1

Any help or advice would be appreciated.
Please let me know if there’s a fix or workaround for this issue. I’m happy to provide additional logs or test a patch.

Thanks so much,
Harpreet

**Comment by zhangwenda0518** [comment]
the same errors happened！

**Comment by oushujun** [comment]
Hello Harpreet,

I tried your mamba installation method, it did not reproduce the issue on my end. You need to make sure LTR_retriever is the laterest version (3.0.2).

Thanks!
Shujun

**Comment by zhangwenda0518** [comment]
thanks！
I ran LTR_retriever (3.0.2) without any error message, but I was running another genome LTR_retriever (3.0.1) and did not encounter the above error. Do I need to run it again？

**Comment by zhangwenda0518** [comment]
I'm sorry, I updated LTR_retriever to 3.0.2 and ran it for a while, but still encountered this error. What should I do

**Comment by zhangwenda0518** [comment]
In my environment, I am running on sample data without any relevant prompts. Should I check my genome?


the genome scaffolds number is 304
the genome contigs number is 839
the longest length is 224921228
the shortest length is 1000
the genome scaffolds size is 2951407848
the genome contig size is 2951354348
the number of N is 53500
the rate of GC is 0.38303796484691
the scaffold N50 is 190645336
the scaffold L50 is 8
the contig N50 is 9730896
the contig L50 is 85
the scaffold N90 is 116512777
the scaffold L90 16
the contig N90 is 2120965
the contig L90 332

**Comment by zhangwenda0518** [comment]
sh: line 1: 1582736 Killed                  timeout -s KILL 300 /home/zhangwenda/mambaforge/envs/EDTA/bin/ltr_finder -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 20 -M 0.85 Scaffold68_sub1 > Scaffold68_sub1.finder.scn
sh: line 1: 1752099 Killed                  timeout -s KILL 300 /home/zhangwenda/mambaforge/envs/EDTA/bin/ltr_finder -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 20 -M 0.85 Scaffold287_sub4 > Scaffold287_sub4.finder.scn
Thread 4 terminated abnormally: Illegal division by zero at /home/zhangwenda/mambaforge/envs/EDTA/share/LTR_retriever/bin/LTR.identifier.pl line 253.
Thread 11 terminated abnormally: Illegal division by zero at /home/zhangwenda/mambaforge/envs/EDTA/share/LTR_retriever/bin/LTR.identifier.pl line 253.
Thread 28 terminated abnormally: Illegal division by zero at /home/zhangwenda/mambaforge/envs/EDTA/share/LTR_retriever/bin/LTR.identifier.pl line 253.
Thread 3 terminated abnormally: Illegal division by zero at /home/zhangwenda/mambaforge/envs/EDTA/share/LTR_retriever/bin/LTR.identifier.pl line 253.
Thread 12 terminated abnormally: Illegal division by zero at /home/zhangwenda/mambaforge/envs/EDTA/share/LTR_retriever/bin/LTR.identifier.pl line 253.

**Comment by kaurharpreet-umn** [comment]
I was able to finish the analyses though.

**Comment by kaurharpreet-umn** [comment]
I think I figured out the issue. It was with my genome fasta file. So, I ran this command

seqkit seq -g genome.fasta > genome.fltd.fasta

My genome file had some spaces in it somewhere. And -g or --remove-gaps option in seqkit package removes those spaces and now everything is working fine. I hope it helps other users as well @zhangwenda0518, if its the same issue. 

Thanks Shujun!

**Comment by Lightoscope** [comment]
I'm getting the same error on a fresh bioconda install.

**Comment by Jnhcau** [comment]
me too
timeout -s KILL 300s /public/home/04034/software/miniconda3/envs/EDTA/bin/gt ltrharvest -index chr1_sub252 -minlenltr 100 -maxlenltr 7000 -mintsd 4 -maxtsd 6 -motif TGCA -motifmis 1 -similar 85 -vic 10 -seed 20 -seqids yes > chr1_sub252.harvest.scn 2> /dev/null

**Comment by ATLord** [comment]
Hi,
I also met the same promble. When I used the same test data, I got the different erros. But all the error is related wite " Killed timeout -s KILL 300s"

Have you found the solution for the problem?
```
Thu Jun 19 09:26:02 CST 2025    EDTA_raw: Check dependencies, prepare working directories.

Thu Jun 19 09:26:17 CST 2025    Start to find LTR candidates.

Thu Jun 19 09:26:17 CST 2025    Identify LTR retrotransposon candidates from scratch.

sh: line 1: 241203 Killed                  timeout -s KILL 300s /data/home/miniconda3/envs/EDTA/bin/gt ltrharvest -index A01_sub22 -minlenltr 100 -maxlenl
sh: line 1: 249640 Killed                  timeout -s KILL 300 /data/home/miniconda3/envs/EDTA/bin/ltr_finder -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 2
sh: line 1: 249725 Killed                  timeout -s KILL 300 /data/home/miniconda3/envs/EDTA/bin/ltr_finder -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 2
sh: line 1: 249727 Killed                  timeout -s KILL 300 /data/home/miniconda3/envs/EDTA/bin/ltr_finder -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 2
sh: line 1: 249743 Killed                  timeout -s KILL 300 /data/home/miniconda3/envs/EDTA/bin/ltr_finder -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 2
sh: line 1: 249772 Killed                  timeout -s KILL 300 `/data/home/miniconda3/envs/EDTA/bin/ltr_finder` -w 2 -C -D 15000 -d 1000 -L 7000 -l 100 -p 2
Thu Jun 19 10:11:15 CST 2025    Finish finding LTR candidates.
```

**Comment by oushujun** [comment]
Hello all,

I catched the bug reported by @Lightoscope too:

> I think I've fixed these issues. The first is just a simple case of a timeout in EDTA_raw.pl in this repository being too short for some circumstances. The second is in LTR.identifier.pl in the LTR_retriever repository whereby several conditional parameters are undefined in edge cases which causes the divide-by-zero issue. I have forked repositories for both with fixes that work for me.

The patched version has been released as v3.0.3. Please check it out! For the timeout kill, you may tolerate these several lines, it's fine if it's just a few. If you don't want any kills, you can increase the timeout from 300 to longer inside the script.

Thanks!
Shujun
